text/0000-global-allocators.md

```diff
@@ -0,0 +1,248 @@
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+#[allocator(allocate_zeroed)]
+pub fn allocate_zeroed(size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Deallocates the memory referenced by `ptr`.
+///
+/// The `ptr` parameter must not be null.
+///
+/// The `old_size` and `align` parameters are the parameters that were used to
+/// create the allocation referenced by `ptr`.
+#[allocator(deallocate)]
+pub fn deallocate(ptr: *mut u8, old_size: usize, align: usize) {
+    ...
+}
+
+/// Resizes the allocation referenced by `ptr` to `size` bytes.
+///
+/// On failure, returns a null pointer and leaves the original allocation
+/// intact.
+///
+/// If the allocation was relocated, the memory at the passed-in pointer is
+/// undefined after the call.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+///
+/// The `old_size` and `align` parameters are the parameters that were used to
+/// create the allocation referenced by `ptr`.
+#[allocator(reallocate)]
+pub fn reallocate(ptr: *mut u8, old_size: usize, size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Resizes the allocation referenced by `ptr` to `size` bytes without moving
+/// it.
+///
+/// The new size of the allocation is returned. This must be at least
+/// `old_size`. The allocation must always remain valid.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+///
+/// The `old_size` and `align` parameters are the parameters that were used to
+/// create the allocation referenced by `ptr`.
+///
+/// This function is optional. The default implementation simply returns
+/// `old_size`.
+#[allocator(reallocate_inplace)]
+pub fn reallocate_inplace(ptr: *mut u8, old_size: usize, size: usize, align: usize) -> usize {
+    ...
+}
+```
+
+Note that `useable_size` has been removed, as it is not used anywhere in the
+standard library.
+
+The allocator functions must be publicly accessible, but can have any name and
+be defined in any module. However, it is recommended to use the names above in
+the crate root to minimize confusion.
+
+An allocator must provide all functions with the exception of
+`reallocate_inplace`. New functions can be added to the API in the future in a
+similar way to `reallocate_inplace`.
+
+## Using an allocator
+
+The functions that an allocator crate defines can be called directly, but most
+usage will happen through the *global allocator* interface located in
+`std::heap`. This module exposes a set of functions identical to those described
+above, but that call into the global allocator. To select the global allocator,
+a crate declares it via an `extern crate` annotated with `#[allocator]`:
```

Comment:
As written, any crate can do this, yeah.

I would be fine restricting allocator selection to the root crate if it simplifies the implementation - I can't think of any strong reasons for needing to select an allocator in a non-root crate.